libp2p/protocols/kademlia/provider.nim

```diff
@@ -206,12 +206,29 @@ proc getProviders*(
 .} =
   ## Get providers for a given `key` from the nodes closest to that `key`.
 
-  var allProviders: HashSet[Provider]
-
-  for chunk in (await kad.findNode(key)).toChunks(kad.config.alpha):
-    let rpcBatch = chunk.mapIt(kad.switch.dispatchGetProviders(it, key))
-    for res in await rpcBatch.collectCompleted(kad.config.timeout):
-      allProviders.incl(res)
+  var
+    nextCandidates: seq[PeerId] = await kad.findNode(key)
+    curCandidates: seq[PeerId]
+    allProviders: HashSet[Provider]
+
+  # run until we find at least K providers
+  var curTry = 0
+  while allProviders.len() < DefaultReplication and curTry < kad.config.retries:
+    curCandidates = nextCandidates
+    nextCandidates = @[]
+    for chunk in curCandidates.toChunks(kad.config.alpha):
+      let rpcBatch = chunk.mapIt(kad.switch.dispatchGetProviders(it, key))
+      for (providers, closerPeers) in await rpcBatch.collectCompleted(
+        kad.config.timeout
+      ):
+        allProviders.incl(providers)
+        for peer in closerPeers:
+          let pid = PeerId.init(peer.id)
+          if not pid.isOk:
+            error "Invalid PeerId getProviders successful reply", peerId = peer.id
+            continue
+          nextCandidates.add(pid.get())
```

Comment:
could there be overlap of `peerIds`? say in current candidates there is `peerA` and `peerB` both returning `peerC` then next loop will request from `peerC` two times?